x/ccv/provider/keeper/key_assignment.go

```diff
@@ -0,0 +1,619 @@
+		// set the mapping from this validator's new consensus address on the consumer
+		// to its consensus address on the provider;
+		// otherwise, the mapping is added when the consumer is registered
+		// note: this state must be deleted through the pruning mechanism;
+		// see ConsumerValidatorsByVscID
+		k.SetValidatorByConsumerAddr(ctx, chainID, consumerAddr, providerAddr)
+
+		// get the previous key assigned for this validator on this consumer chain
+		oldConsumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr)
+		if found {
+			// mark this old consumer key as prunable once the VSCMaturedPacket
+			// for the current VSC ID is received;
+			// note: this state is removed on receiving the VSCMaturedPacket
+			k.AppendConsumerAddrsToPrune(
+				ctx,
+				chainID,
+				k.GetValidatorSetUpdateId(ctx),
+				utils.TMCryptoPublicKeyToConsAddr(oldConsumerKey),
+			)
+		} else {
+			// the validator had no key assigned on this consumer chain
+			providerKey, err := validator.TmConsPublicKey()
+			if err != nil {
+				return err
+			}
+			oldConsumerKey = providerKey
+		}
+
+		// check whether the validator is valid, i.e., its power is positive
+		oldPower := k.stakingKeeper.GetLastValidatorPower(ctx, sdk.ValAddress(validator.OperatorAddress))
+		if oldPower > 0 {
+			// to enable multiple calls of AssignConsumerKey in the same block by the same validator
+			// the pending key assignment should not be overwritten
+			if _, found := k.GetPendingKeyAssignment(ctx, chainID, providerAddr); !found {
+				// store old key and power for modifying the valset update in EndBlock;
+				// note: this state is deleted at the end of the block
+				k.SetPendingKeyAssignment(
+					ctx,
+					chainID,
+					providerAddr,
+					abci.ValidatorUpdate{PubKey: oldConsumerKey, Power: oldPower},
+				)
+			}
+		}
+	}
+
+	// set the mapping from this validator's provider address to the new consumer key;
+	// overwrite if already exists
+	// note: this state is deleted when the validator is removed from the staking module
+	k.SetValidatorConsumerPubKey(ctx, chainID, providerAddr, consumerKey)
+
+	return nil
+}
+
+// ApplyKeyAssignmentToInitialValset applies the key assignment to the initial validator
+// for a consumer chain
+func (k Keeper) ApplyKeyAssignmentToInitialValset(
+	ctx sdk.Context,
+	chainID string,
+	updates []abci.ValidatorUpdate,
+) (newUpdates []abci.ValidatorUpdate) {
+	newUpdates = updates
+	for i, update := range updates {
+		providerAddr := utils.TMCryptoPublicKeyToConsAddr(update.PubKey)
+		if consumerKey, found := k.GetValidatorConsumerPubKey(ctx, chainID, providerAddr); found {
+			newUpdates[i].PubKey = consumerKey
+			consumerAddr := utils.TMCryptoPublicKeyToConsAddr(consumerKey)
+			// set the mapping from this validator's new consensus address on the consumer
+			// to its consensus address on the provider
+			k.SetValidatorByConsumerAddr(ctx, chainID, consumerAddr, providerAddr)
+		}
+	}
+	return newUpdates
```

Comment:
I think the difference is the call `k.SetValidatorByConsumerAddr(ctx, chainID, consumerAddr, providerAddr)`

I think there's a bug. Consider the following

A validator `V` assigns a consumer key `K` for a consumer chain which is not yet registered.
The initial validator set, when the consumer chain genesis is created does not contain `V`. 
The slash lookup assignment `K -> V` is never written.
A slash lookup returns `K->K` because `K->V` was never explicitly written.
`K` points to nothing so the validator can't be slashed.